### PR TITLE
Extension versioning

### PR DESCRIPTION
**Pull request by mathbagu**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#28 

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
The extension are always read/write in the current version

**What is the new behavior (if this is a feature change)?**
This version prepare the support of the versioning in the extension. An extension can now have several XML serialization depending on the version.
We also refactor the IIDM tests to separate the test of the grid model and the test relative to the XIIDM serialization. We also introduce a new extension library with 3 extensions for unit testing.

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: include/powsybl/iidm/converter/xml/AbstractExtensionXmlSerializer.hpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (sebalaig) [inline include/powsybl/iidm/converter/xml/AbstractExtensionXmlSerializer.hpp:2]: 2020 ;)

File: include/powsybl/iidm/converter/xml/ExtensionXmlSerializer.hpp [hunk 2]
```diff
@@ -28,35 +30,98 @@ namespace xml {
 class NetworkXmlReaderContext;
 class NetworkXmlWriterContext;
 
+/**
+ * An ExtensionProvider able to serialize/deserialize extensions from XML
```
Comment (sebalaig) [inline include/powsybl/iidm/converter/xml/ExtensionXmlSerializer.hpp:34]: You should use {@link ExtensionProvider}, no ?

File: include/powsybl/iidm/converter/xml/ExtensionXmlSerializer.hpp [hunk 3]
```diff
@@ -28,35 +30,98 @@ namespace xml {
...
+     * @param extension The extension to write to XML
+     * @param context The current XML context
+     */
     virtual void write(const Extension& extension, NetworkXmlWriterContext& context) const = 0;
 
 protected:
+    /**
+     * A utility function to cast a generic extension into its specific type E
+     *
+     * @tparam E The expected type of the extension
```
Comment (sebalaig) [inline include/powsybl/iidm/converter/xml/ExtensionXmlSerializer.hpp:113]: typo `tparam`

File: src/iidm/converter/xml/AbstractVersionableExtensionXmlSerializer.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    return m_extensionVersions.at(IidmXmlVersion::CURRENT_IIDM_XML_VERSION().toString(".")).back();
+}
+
+}  // namespace xml
+
+}  // namespace converter
+
+}  // namespace iidm
+
+}  // namespace powsybl
```
Comment (sebalaig) [inline src/iidm/converter/xml/AbstractVersionableExtensionXmlSerializer.cpp:67]: Add a new line

File: test/iidm/converter/xml/IdentifiableExtensionXmlSerializerTest.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace powsybl {
+
+namespace iidm {
+
+namespace converter {
+
+namespace xml {
+
+BOOST_AUTO_TEST_SUITE(IdentfiableExtensionXml)
```
Comment (sebalaig) [inline test/iidm/converter/xml/IdentifiableExtensionXmlSerializerTest.cpp:27]: Typo `IdentfiableExtensionXml`

File: test/iidm/converter/xml/extensions/TerminalMockExt.hpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    const Terminal& getTerminal() const;
+
+    TerminalMockExt& setTerminal(const Terminal& terminal);
+
+private:  // Extension
+    void assertExtendable(const stdcxx::Reference<Extendable>& extendable) const override;
+
+private:
+    stdcxx::CReference<Terminal> m_terminal;
```
Comment (sebalaig) [inline test/iidm/converter/xml/extensions/TerminalMockExt.hpp:47]: It looks like a `std::reference_wrapper` can be used instead.

File: test/iidm/converter/xml/extensions/TerminalMockExt.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+}  // namespace extensions
+
+}  // namespace xml
+
+}  // namespace converter
+
+}  // namespace iidm
+
+}  // namespace powsybl
+#endif  // POWSYBL_IIDM_CONVERTER_XML_EXTENSIONS_TERMINALMOCKEXT_HPP
```
Comment (sebalaig) [inline test/iidm/converter/xml/extensions/TerminalMockExt.hpp:59]: Add a new line before this line

File: test/iidm/converter/xml/xiidm.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (sebalaig) [inline test/iidm/converter/xml/xiidm.cpp:2]: 2020

File: include/powsybl/iidm/converter/xml/AbstractExtensionXmlSerializer.hpp (hunk 1 above)
Comment (mathbagu) [reply include/powsybl/iidm/converter/xml/AbstractExtensionXmlSerializer.hpp:2]: Git consider this file as a new file whereas I just move it. I will fix it

File: include/powsybl/iidm/converter/xml/ExtensionXmlSerializer.hpp (hunk 2 above)
Comment (mathbagu) [reply include/powsybl/iidm/converter/xml/ExtensionXmlSerializer.hpp:34]: Javadoc looks ok: I will keep it like this

**Pull request merged**